src/modules/kafka/kfk.c

```diff
@@ -0,0 +1,1175 @@
+		rd_kafka_metadata_topic_t *t = &metadatap->topics[i];
+		if (t->topic) {
+			LM_DBG("Metadata Topic: %s\n", t->topic);
+			if (strncmp(topic_name->s, t->topic, topic_name->len) == 0) {
+				topic_found = 1;
+				LM_DBG("Metadata Topic (%s) found!\n", t->topic);
+				break;
+			}
+		}
+	} // for (i=0; i<m->topic_cnt; i++)
+
+	/* Destroy metadata. */
+	rd_kafka_metadata_destroy(metadatap);
+
+	if (topic_found == 0) {
+		LM_DBG("Topic not found: %.*s\n", topic_name->len, topic_name->s);
+		return 0;
+	}
+
+	LM_DBG("Topic found: %.*s\n", topic_name->len, topic_name->s);
+	return 1;
+	
+error:
+
+	/* Destroy metadata. */
+	if (metadatap) {
+		rd_kafka_metadata_destroy(metadatap);
+	}
+	
+	return -1;
+}
+
+/**
+ * \brief get a topic based on its name.
+ *
+ * \return topic if it founds a matching one, NULL otherwise.
+ */
+static rd_kafka_topic_t* kfk_topic_get(str *topic_name)
+{
+	rd_kafka_topic_t *result = NULL;
+
+	if (!topic_name || topic_name->len == 0 || topic_name->s == NULL) {
+		LM_ERR("Bad topic name\n");
+		goto clean;
+	}
+
+	kfk_topic_t *ktopic = kfk_topic;
+	while (ktopic) {
+		kfk_topic_t *next = ktopic->next;
+
+		if (topic_name->len == ktopic->topic_name->len &&
+			strncmp(topic_name->s, ktopic->topic_name->s, topic_name->len) == 0) {
+			LM_DBG("Topic name match: %.*s\n",
+				   ktopic->topic_name->len,
+				   ktopic->topic_name->s);
+			result = ktopic->rd_topic;
+			break;
+		}
+
+		ktopic = next;
+	}
+	
+clean:
+	
+	return result;
+}
+
+/**
+ * \brief send a message to a topic.
+ *
+ * \param topic_name name of the topic
+ * \param message message to send.
+ *
+ * \return 0 on success.
+ */
+int kfk_message_send(str *topic_name, str *message)
+{
+    /* Get topic from name. */
+	rd_kafka_topic_t *rkt = kfk_topic_get(topic_name);
+
+	if (!rkt) {
+		LM_ERR("Topic not found: %.*s\n", topic_name->len, topic_name->s);
+		return -1;
+	}
+
+	/* Send a message. */
+	if (rd_kafka_produce(
+			rkt,
+			RD_KAFKA_PARTITION_UA,
+			RD_KAFKA_MSG_F_COPY,
+			/* Payload and length */
+			message->s,
+			message->len,
+			/* Optional key and its length */
+			NULL, 0,
+			/* Message opaque, provided in
+			 * delivery report callback as
+			 * msg_opaque. */
+			NULL) == -1) {
+		rd_kafka_resp_err_t err = rd_kafka_last_error();
+		LM_ERR("Error sending message: %s\n", rd_kafka_err2str(err));
+
+		return -1;
+	}
+
+	LM_DBG("Message sent\n");
+
+	/* Poll to handle delivery reports */
+	rd_kafka_poll(rk, 0);
+	LM_DBG("Message polled\n");
+
+	return 0;
+}
+
+/**
+ * \brief Initialize statistics.
+ *
+ * \return 0 on success.
+ */
+int kfk_stats_init()
+{
+	LM_DBG("Initializing statistics\n");
+
+	stats_lock = lock_alloc();
+	if (!stats_lock) {
+		LM_ERR("Cannot allocate stats lock\n");
+		return -1;
+	}
+
+	if(lock_init(stats_lock) == NULL) {
+		LM_ERR("cannot init stats lock\n");
+		lock_dealloc(stats_lock);
+		stats_lock = NULL;
+		return -1;
+	}
+
+	stats_general = shm_malloc(sizeof(kfk_stats_t));
```

Comment:
Where you shm_free this stats_general memory? Or is this deleted together with the other elements?

Reply:
Hi Henning!
stats_general memory is freed in kfk_stats_close function node by node by kfk_stats_topic_free.

Reply:
Thank you